### PR TITLE
Issue 56- New feature- Countdown Timer with new helpers, tests, and code coverage.

### PR DESCRIPTION
**Pull request by kaxcode**

### CODE REVIEW

File: src/Timer.test.js [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+  beforeEach(() => {
+    wrapper = shallow(<Timer />);
+  });
+
+  it('renders correctly', () => {
+    const tree = renderer.create(wrapper).toJSON();
+    expect(tree).toMatchSnapshot();
+  });
+
+  describe('when the `+/-` buttons are clicked', () => {
```
Comment (dlindahl) [inline src/Timer.test.js]: "`clicked`" in the test name is redundant. When these tests are run, they generate a test statement. So in the first test, it generates a test name of:

`Timer when the '+/-' buttons are clicked the '+' hours button is clicked should update the state timeRemaining to increase by 1`

That is not a very easy to understand statement.

Consider the following:

```js
describe('Timer', () => {
  describe('Increment Buttons', () => {
    describe('clicking the + button', () => {
      it('increases the timeRemaining state by 1')
    });
  })
})
```

The catch here is that the "`the `+` hours button is clicked`" `describe` block isn't really doing anything for us other than adding unnecessary indentation to the test which can make it harder to read.

You could change that by adding a `beforeEach` test that encapsulates all of the "clicking" logic for all the subsequent test. But given the nature of these tests, that would be unnecessary.

I recommend simplifying the tests as so:

```
describe('Timer', () => {
  describe('clicking the hour increment button', () => {
    it('increases the time remaining')
    it('prevents a time remaining increase above the max time allowed')
  })

  describe('clicking the hour decrement button', () => {
    it('decreases the time remaining')
    it('prevents a time remaining decrease below the min time allowed')
  })
})
```

These tests are concise in their description of the test scenario and are assertive as to what the expectation is. When run, they generate the test statement `Timer clicking the hour increment button increases the time remaining` which is much easier to read.

Also of note, I removed the `by 1`/`above 23` statements for two reasons:
1) The tests are not actually asserting that the state is increasing by 1. One is testing an increase of three by clicking three times and another is testing the decrease of 2 by clicking two times.

2) By including a specific value in the test name, its prone to becoming out of date if the business logic changes. Much like the following comment:

```js
// Results are always sorted in a descending order
const results = db.query('select * from theTable order ASC')
```

Not only does the comment not really tell us anything useful, but its ultimately unnecessary because the code already states the sort order. So when the business logic inevitably changes, it must be made twice or risk leaving a confusing code comment.

Bringing this back to tests, the test name should stay focused on describing what is being tested and not the implementation details. This keeps things concise and less prone to being in error when a change occurs. Let the assertions dictate what the business logic success state should be and keep the test name relatively high-level.

File: src/Timer.test.js [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+        button.simulate('click');
+
+        // Assert
+        expect(wrapper.state().timeRemaining).toEqual(3*hours);
+        expect(wrapper.state().timeRemaining).not.toBe(1*hours);
+      });
+
+      it('should not let hours increase above 23', () => {
+        // Arrange
+        const button = wrapper.find('.hours .plus-btn').first();
```
Comment (dlindahl) [inline src/Timer.test.js]: Its not a great idea to couple tests to class names since those can change frequently. I'd recommend adding a `data-test-id` or even *gasp* and `id` attribute (since there is only one "increase hour" button).

As an aside, I'm surprised this selector is working since its selector the `.plus-btn` component that has a parent of `.hours`. The actual tree says that the `.plus-btn` and `.hours` component are one in the same.

In other words, I would expect the selector to be `.hours.plus-btn`.

It is probably worth taking a look at because you may not be testing what you think you are, the application state is ambiguous to the difference in the two selectors, or Jest has a bug!

File: src/Timer.test.js [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+        button.simulate('click');
+
+        // Assert
+        expect(wrapper.state().timeRemaining).toEqual(0);
+        expect(wrapper.state().timeRemaining).not.toBe(-1);
+      });
+    });
+  });
+
+  jest.useFakeTimers();
```
Comment (dlindahl) [inline src/Timer.test.js]: This looks misplaced. The Jest docs seem to want it at the top of the file before the very first `describe` block.

File: src/Timer.test.js [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+
+  describe('when the `Start` button is clicked', () => {
```
Comment (dlindahl) [inline src/Timer.test.js]: nit: Try to avoid using quotation marks in test descriptions. You'll save yourself a lot of annoyance.

File: src/Timer.test.js [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+      startButton.simulate('click');
+
+      // Assert
+      expect(setInterval).toHaveBeenCalledTimes(1);
+      expect(setInterval).toHaveBeenLastCalledWith(expect.any(Function), 1000);
+    });
+  });
+
+  describe('when the `Pause` button is clicked', () => {
+    it('should pause the ', () => {
```
Comment (dlindahl) [inline src/Timer.test.js]: It should pause the... what?

File: src/Timer.test.js [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+      // Act
+      clearButton.simulate('click');
+
+      // Assert
+      expect(wrapper.state().timeRemaining).toEqual(0);
+      expect(wrapper.state().timeRemaining).not.toBe(23*hours);
+    });
+  });
+
+  describe('when the `tick` is called', () => {
```
Comment (dlindahl) [inline src/Timer.test.js]: nit: When setting up the test of a specific function, use the following syntax:

```js
describe('Timer', () => {
  describe('#tick', () => {
    it('decreases the time remaining')
  });
})
```

This will generate a test statement of `Timer#tick decreases the time remaining`.

File: src/utils/humanizeTimer.js [hunk 7]
```diff
@@ -11,3 +11,15 @@ export function millisecondsToHuman(time) {
     centisecond.toString().padStart(2, '0')
   ].join(':');
 }
+
+export function secondsToHour(time) {
```
Comment (dlindahl) [inline src/utils/humanizeTimer.js]: Test this

File: src/utils/humanizeTimer.js [hunk 8]
```diff
@@ -11,3 +11,15 @@ export function millisecondsToHuman(time) {
...
+  return Math.floor(time / 60 / 60);
+}
+
+export function secondsToMinutes(time) {
```
Comment (dlindahl) [inline src/utils/humanizeTimer.js]: Test this

File: src/utils/humanizeTimer.js [hunk 9]
```diff
@@ -11,3 +11,15 @@ export function millisecondsToHuman(time) {
...
+  return Math.floor(time / 60 % 60);
+}
+
+export function seconds(time) {
```
Comment (dlindahl) [inline src/utils/humanizeTimer.js]: Test this

File: package.json [hunk 10]
```diff
@@ -23,7 +23,7 @@
     "build": "react-scripts build",
     "test:lint:js": "eslint --ext=js --ext=jsx .",
     "test:unit": "react-scripts test --env=jsdom",
-    "test": "react-scripts test --env=jsdom",
+    "test": "react-scripts test --coverage --env=jsdom",
```
[the remark this answers is not in the document]
Comment (dlindahl) [reply package.json]: Since we are here, `test` is now 99% the same as `test:unit`. And when `npm test` is run, `test:lint:js` does not get run at all.

Do the following instead:

```json
"scripts": {
  "lint": "npm-run-all --parallel lint:*",
  "lint:js": "eslint --ext=js --ext=jsx .",
  "pretest": "npm run lint",
  "test:unit": "react-scripts test --env=jsdom",
  "test": "npm-run-all --parallel test:*"
}
```

This groups all the linting behavior into a single, runnable namespace. Same with all of the test behaviors. When `npm test` is run, the `pretest` hook will kick in, running all of the lint tasks and, if there are no issues, run the entire test suite.

Admittedly this does have some YAGNI in it in that we have no other linting to do other than JS and no other tests other than unit, but its a good place to point out this pattern.

File: src/Timer.test.js (hunk 6 above)
Comment (kaxcode) [reply src/Timer.test.js]: I already have the `describe('Timer', ()`. Updated everything else.

File: src/__snapshots__/Stopwatch.test.js.snap [hunk 11]
```diff
@@ -1,64 +1,529 @@
 // Jest Snapshot v1, https://goo.gl/fbAQLP
 
 exports[`renders correctly 1`] = `
-<div
-  className="row"
+ShallowWrapper {
```
Comment (dlindahl) [inline src/__snapshots__/Stopwatch.test.js.snap]: Its good that the `Each child in an array or iterator should have a unique "key" prop` error is now resolved, but the result is that the snapshot file has exploded from 65 lines to more than 500 for one test. You and @meingorn should get together to discuss if this is OK and if not, what you should do about it.

File: src/Timer.test.js (hunk 2 above)
Comment (kaxcode) [reply src/Timer.test.js]: I used id attributes on the buttons.

File: package.json (hunk 10 above)
[the remark this answers is not in the document]
Comment (kaxcode) [reply package.json]: We'll fix this next sprint.

File: src/utils/humanizeTimer.js (hunk 9 above)
Comment (kaxcode) [reply src/utils/humanizeTimer.js]: Check humanizeTimer.test.js

File: src/utils/humanizeTimer.js (hunk 8 above)
Comment (kaxcode) [reply src/utils/humanizeTimer.js]: Check humanizeTimer.test.js

File: src/utils/humanizeTimer.js (hunk 7 above)
Comment (kaxcode) [reply src/utils/humanizeTimer.js]: Check humanizeTimer.test.js

File: src/Timer.test.js (hunk 2 above)
Comment (dlindahl) [reply src/Timer.test.js]: Did you find out why the selector works as `.hours .plus-btn` _as well as_ `.hours.plus-btn`? That is weird behavior and should not work in both cases.

File: src/Timer.js [hunk 12]
```diff
@@ -1,11 +1,87 @@
...
+    }
+  };
+
+  decreaseSeconds = () => {
+    if (this.state.timeRemaining > 0) {
+      this.setState({ timeRemaining: (this.state.timeRemaining - this.SECONDS ) });
+    }
+  };
+
+  time = Object.assign({}, this.state.timeRemaining);
```
[the remark this answers is not in the document]
Comment (kaxcode) [reply src/Timer.js]: removed line

File: src/utils/humanizeTimer.test.js [hunk 13]
```diff
@@ -0,0 +1,43 @@
+import { secondsToHour, secondsToMinutes, seconds } from './humanizeTimer';
+
+describe('Timer helpers', () => {
+  it('it should return seconds in hours', () => {
```
Comment (dlindahl) [inline src/utils/humanizeTimer.test.js]: Lots of `its` here and its missing a `describe` block for the function under test. Try something like this:

```js
describe('humanizeTimer', () => {
  describe('#secondsToHour', () => {
    it('converts seconds to a rounded-down whole hour', () => {
      // assert 1800 == 0
      // assert 3600 == 1
      // assert 5400 == 1
    })
  })
})
```

While having one assertion per test is generally considered a best practice, its not an absolute rule. In the name of clarity, I think its OK to put several assertions that test that varied inputs all prove the test statement to be true. Breaking each test scenario into unique test cases doesn't really gain us much in this particular scenario.

If all of this sounds subjective, you're right. But its all about striking a balance between under-communicating and over-communicating.

Now, if this function handled specific inputs in a special way, then it would make sense to put that into its own test scenario:

```js
it('converts negative seconds to 0', () => {
  // assert -3600 == 0
})
```

This scenario isn't just about rounding, so you would probably want to explicitly test and document that behavior.

**Comment by dlindahl** [comment]
This looks good now. The only thing left is to squash your commits so that they make sense in regards to the issue that you are fixing.

For example, the linting task you changed should be its own commit since it has nothing to do with #56. Anything related to actually closing #56 should be in one commit.

Also, the name of this PR is... not great. Please change it to something more descriptive of what this PR actually does.
